### ISSUE TITLE
mat: triangular matrices don't respect the .mat.Diag field

### ISSUE THREAD

**Issue by kortschak** [gonum/gonum#692]
The `.mat.Diag` field of the triangular types is used by nearly none of the operations in mat (one internal method checks it to panic and the new diagonal extraction method on `Diagonal` allows unit diagonals to be obtained (FWIW). However indexing a `TriDense` for example does not check `t.mat.Diag`. This should probably be done (although there is currently no way for a user to set the state to unit, that may happen).

**Comment by btracey** [comment]
I don't think there's any pressing need to expose the `Diag` field as part of the interface. The reason to have the field is so you can put two triangles together in one Dense, or to put a triangular and a symmetric together. That's a big part of why BLAS works, but it's somewhat counter the spirit of `mat`. For instance, it's impossible in general to maintain the `Diag` property when the `TriDense` is a receiver for most operations. That said, we support `RawTriangular`, and so we should respect the full interface. This would also enable a user to have a type for this kind of storage (that implements both `RawTriangular` and `RawSymmetric` for instance.

If that's the policy, presumably that also means we should be supporting `Lower` symmetric matrices as well?

**Comment by kortschak** [comment]
The issue here is to ensure that the check is done correctly if the `Diag` field is set differently to how the package corrently does by default. As I wrote above, this is not urgent since there is no safe way for a user to do that contrary setting of `Diag` since we don't have a raw setter for these types. However, when we do, this will have to happen.

BTW, storing two triangles|symmetics in a dense is potentially hazardous if one is a parameter and one is a receiver since that will fail a shadow check.

**Comment by kortschak** [comment]
By "receiver" what do you mean?

The reason for having a raw setter is to allow users to fill in areas of BLAS API that we don't cover. This was an integral part of the API design from the start. Shadow detection prevents egregiously unsafe uses, but there are safe uses that are made available by being able to do this.

Note that with a raw setter, once the `blas64.Triangular` or `blas64.TriangularBand` has been handed in, the `Diag` field is not mutable externally, so the maintenance issue is not a problem.

**Comment by btracey** [comment]
By receiver, I mean something like
```
t.FromRawTriangular(raw)
t.AddTri(t1, t2)
```
It is impossible to maintain `t` as a Diag in this case. If we want to support `Diag=true`, then we have to add in many panic checks. I do not think we want to go the route of "unless t1 and t2 have zeros on the diagonal"

**Comment by kortschak** [comment]
Sure. All  mutating operations on triagular types can be documented to cause a panic when the value is `blas.Unit`.

**Comment by btracey** [comment]
... which doesn't really work if we're going to do computations on the `Raw` types directly. 

Do we really need the `RawXxx` types? Could we not instead replace them

```
type `DenseViewer` interface{
     DenseViewer() *Dense
}

func (m *Dense) FromBlasGeneral(b blas64.General) {
    // stuff
}
```

I believe this enables the same set of optimizations. If a type can currently implement `RawMatrix`, they can implement `DenseViewer`, as
```
func (m MyType) DenseViewer() {
    b := m.RawMatrix()
    var d mat.Dense
     return &d.FromBlasGeneral(b)
}
```
For the other types (i.e. FromBlasSymmetric), we enforce our internal invariants.

The benefit is that internally we only need to think about `Dense`, `SymDense`, etc. and do not need to worry about the extra fields. We can successfully maintain our internal invariants, and we get to simplify our internal code, because now we only have to think about our Dense types.

**Comment by btracey** [comment]
I'm not suggesting we remove the Raw methods from the Dense types. I agree with you there. I am instead suggesting that we make implementing `DenseViewer` etc. the "official" way to get fast-pathing. Currently if you implement `RawTriangular` (for example) you get fast-pathing, but also the wrong answer in many cases. I think my suggestion makes it clearer what the contract is, and makes it easier to implement fast pathing.

**Comment by btracey** [comment]
(the wrong answer if you have Diag for instance)

**Comment by btracey** [comment]
Enforcing those checks when asserting if a `RawXxer` will have cost at the call site, I agree that it will not if we force use through the `Set` methods.

**Comment by kortschak** [comment]
Yeah, I'm not suggesting check for `Raw` calls, that's too late.

How does your approach above handle non-general type optimisations? I think I am missing something here.

**Comment by btracey** [comment]
I don't think I understand the question. I am proposing that instead of checking for `RawMatrix`, we check for `DenseViewer`.  This would handle types which are not `*Dense` itself by having a `FromRawMatrix` method on `*Dense`, a la my comment above.

In terms of the internal optimizations, the code would be much simpler and easier to read if we could just check for `*Dense` directly. I'm not totally clear what the use-case is for alternate implementations, but assuming it's something along the lines of  1) Do some stuff 2) Make an optimized call into a `mat` function 3) Do some more stuff , then casting into a `*Dense` between 1 and 2 would not be that onerous, and the benefit is to hugely improve the internal code (I think).

**Comment by btracey** [comment]
Similarly, we would expose a `FromBlasTriangular`  on `*TriDense`, and then in the internal code check for a `*TriDense` instead of a `RawTriangler`

**Comment by kortschak** [comment]
Yeah, I understand the shape of the change, I just don't understand why use `*Dense` instead of `blas64.General` (which will be the type that actually is worked on) and I don't understand how we then signal BLAS optimisations dependent on the arguments' types (e.g. symmetric or triangular or banded) that would allow optimisations to the BLAS calls being made.

If you could sketch it out for me, maybe I'd understand better.

**Comment by btracey** [comment]
(I agree this is a reasonable request, haven't found the time yet)
